### ISSUE TITLE
"shell out" or filter though bash command?

### ISSUE THREAD

**Issue by ghost** [stedolan/jq#147]
Given this file

``` json
{
  "releases": [
    {
      "date": "1998-05-12"
    },
    {
      "date": "1997-05-12"
    },
    {
      "date": "1999-05-12"
    }
  ]
}
```

I would like to run the `date` values through the `date` command, example

``` bash
$ date +%s -d 1998-05-12
894949200
```

so that the final result is

``` json
{
  "releases": [
    {
      "date": 894949200
    },
    {
      "date": 863413200
    },
    {
      "date": 926485200
    }
  ]
}
```

Is something like this possible?

**Comment by jrdriscoll** [comment]
Off and on I end up doing work with big piles of unstructured data that I need to make sense of, and I traditionally have used giant shell pipelines to sort out the needles in the haystacks.  Lately I have being doing more and more with jq, primarily because it is much less error prone due to its clean (but occasionally surprising) semantics.  This feature request is one of the most frequent reasons I have to "drop out" of jq to process some data.

In the traditional unix-pipeline-awk world, you would would use awk's `system()` call with a string shell command constructed from the record you are processing (e.g.` system("date -d @" $2)`).  Based on my (possibly idiosyncratic) usage of jq, I think this would fit reasonably well into the jq world:

    jq ' .date = system("date -d @" + (.epoch | tostring)) '

Note that the json string argument will need to be converted to raw, and in the example we will probably want the output to be converted to a json string.  But the result could have been json.  I expect the user will want some control over how the input/output conversions are handled.

If date returned more than one result (although I don't think it will) this also seems to work fine.

So `system()` will work find for date, which processes a single date at a time, but what if you want to NFKC normalize some strings in a few million json records?  In awk, you would solve this using a "coprocess"  to which you can both read and write.  The syntax is awkward, and derived from the Korn shells `|&` operator.

    awk '
      BEGIN {
        uconv = "uconv -b 1 -f utf-8 -t utf-8 -x \"::nfkc;\"";
        PROCINFO[uconv, "PTY"] = 1;
      }
      {
        print $2 |& uconv;
        uconv |& getline normalized;
       ...
      }
      END {
        close(uconv);
      }
    '
In these situations, deadlock is a possibility and that is why the example asks awk to use a pty for the coprocess.  I left out the usual `stdbuf` incantations that try to force the coprocess to work unbuffered.

The low level reading and writing is quite flexible, but not so clean.  If you constrain it to be a filter that takes one input and produced one output, It looks more like a jq filter and seems to fit in nicely, although as a filter it will not produce an interesting json output unless the coprocess does.

**Comment by jrdriscoll** [comment]
I have problems like that of the OP all the time where I need to get data out of one system, transform it, and then somehow join it back in with the original data.  With the newish `input` it is *much* easier to do this sort of thing with jq.  Here is a reasonably clean way to do way the OP poster wants (admittedly, my tolerance "reasonable" and "clean" in these matters may not be representative):

```#!/bin/bash

# need objects on single line for subsequent paste to work
# paste is effectively joining on implicit key = line_number

cat releases.json \
 | jq -c '.' >r-c.json 

# convert contained dates to pipe separated string in jq
# convert to space separated epochs in awk
# paste json objects, corresponding epochs onto single line
# read the epochs after each object with input
# note that the test for type not string also removes nulls

cat r-c.json \
  | jq -r ' 
    [ .releases[].date ] 
    | map(if (.|type)!="string" then empty else . end) 
    | reduce .[] as $d (""; . + $d + "|") 
  ' \
  | gawk -F\| ' { 
    for (i=1; i<NF; i++) {
      "date +%s -d " $(i) | getline d;
      printf d " "; 
    } 
    printf "\n";
  } ' \
  | paste -d\  r-c.json - \
  | jq ' 
    .releases = ( .releases 
      | map(if (.date|type)=="string" then .date=input else . end)
    ) 
  '
```

**Comment by nicowilliams** [comment]
Yes, I resort to these kinds of tricks too.  A shell-out should probably be a high priority.  I may even work on it this coming weekend, we'll see.

**Comment by jrdriscoll** [comment]
Well, while I do think that a "shell-out" feature would help many users
of jq, my point was actually that "input" solved essentially ALL my
problems (although not in the most terse or elegant way).  And it is
true you need to be a relatively sophisticated user to do so.

**Comment by nicowilliams** [comment]
Yes, `input` and `inputs` solved and/or help work around a number of problems.  I'm quite happy about how `input` and `inputs` turned out.

**Comment by svnpenn** [comment]
@nicowilliams I should mention I do not think this feature is required anymore in this specific case. I have not had time to check, but I believe the new JQ date commands can be used to fix my issue here. However others might still be interested in generic JQ `system` command.

**Comment by nicowilliams** [comment]
@svnpenn Right, for datetime-related tasks a shell-out is not needed.  You'll note that we added the sorts of things we needed that we re relatively easy to add :)

A shell-out wouldn't be so hard to code, but first we needed to work out a privilege management model that would work for that and I/O in general.

I see two shell-out forms:  `CMD | popen` and `CMD | popen(inputs_for_cmd)`.  The former would read from the command (it would map to popen() with "r"), and the latter would write to the command (it would map to popen() with "w").

**Comment by nicowilliams** [comment]
See #1005.

**Comment by svnpenn** [comment]
https://github.com/stedolan/jq/issues/925

**Comment by svnpenn** [comment]
https://github.com/stedolan/jq/issues/503

**Comment by svnpenn** [comment]
https://github.com/stedolan/jq/issues/255

**Comment by mterron** [comment]
I'm converting binary hashes that has been base64 encoded back into hexadecimal representation. @base64d breaks when the decoded value is not a UTF string.

I want to be able to "shell-out" to execute `base64 -d | xxd -p -c32 | tr -cd '[:alnum:]\n'` on a field. The output of that pipeline is an hexadecimal encoded hash that is a valid json value. At the moment I have an `awk` hack to do it.

**Comment by cup** [comment]
@mterron have you considered using a proper programming language? i know it
might be daunting, but i might be able to help if you have some sample data.
here are some links:

Python
------

- https://docs.python.org/library/json#json.loads
- https://docs.python.org/library/base64#base64.b64decode

Ruby
-----

- http://ruby-doc.org/stdlib/libdoc/json/rdoc/JSON.html#method-i-parse
- http://ruby-doc.org/stdlib/libdoc/base64/rdoc/Base64.html#method-i-decode64

**Comment by mterron** [comment]
I have but `jq` does 99.9% of what I want so why bother. I hacked that awk thing in 10 minutes, it'd take me 10 times as much to do it in Python or Ruby and also that adds a huge dependency framework to my pipeline that I'd rather not have.


Example json input (after a lot of manipulation with `jq`):

```json
{
  "component": "1000hz-bootstrap-validator",
  "version": "0.10.2",
  "hashes": [
    {
      "file": "validator.js",
      "base64": "sha256-eXmycr1Eg/2vbFxVvM6avBWolBl8B6VkTRV5/9B9kto="
    },
    {
      "file": "validator.min.js",
      "base64": "sha256-mbv8R/8RTicMdfYPxNwD4QVAvNPO6Ht+ZDW9EK0gNHM="
    }
  ]
}
```
and json output:
```json
{
  "component": "1000hz-bootstrap-validator",
  "version": "0.10.2",
  "hashes": [
    {
      "file": "validator.js",
      "base64": "sha256-eXmycr1Eg/2vbFxVvM6avBWolBl8B6VkTRV5/9B9kto=",
      "sha256": "7979b272bd4483fdaf6c5c55bcce9abc15a894197c07a5644d1579ffd07d92da"
    },
    {
      "file": "validator.min.js",
      "base64": "sha256-mbv8R/8RTicMdfYPxNwD4QVAvNPO6Ht+ZDW9EK0gNHM=",
      "sha256": "99bbfc47ff114e270c75f60fc4dc03e10540bcd3cee87b7e6435bd10ad203473"
    }
  ]
}
```

**Comment by cup** [comment]
~~~
$ echo sha256-eXmycr1Eg/2vbFxVvM6avBWolBl8B6VkTRV5/9B9kto= | base64 -d
□□base64: invalid input
~~~

**Comment by mterron** [comment]
You need to process the string first,  `sub()` is nice.

```bash
$ echo "eXmycr1Eg/2vbFxVvM6avBWolBl8B6VkTRV5/9B9kto=" | base64 -d | xxd -p -c32
7979b272bd4483fdaf6c5c55bcce9abc15a894197c07a5644d1579ffd07d92da
```

**Comment by cup** [comment]
Sorry, I see you are moving goalposts. I have dealt with that before and I have no tolerance policy. Good luck.

**Comment by mterron** [comment]
Meaning? I just provided a real life use case for the shell-out feature. What I want to do can't be done with `jq`  without a `shell-out` feature.

It is an escape hatch.

**Comment by cup** [comment]
https://meta.stackoverflow.com/questions/251758/-/252077

**Comment by mterron** [comment]
I'm not sure what you are trying to say tbh.

**Comment by wtlangford** [comment]
This conversation is getting a little unnecessarily heated, so let's please remain polite and civil, shall we?

On topic, though: @mterron The ability to shell out is something we're working on, and there are some branches floating around with the capability. They're a little buggy at the moment, and may not actually have direct support for shelling out yet (I'd have to check), but they contain the necessary groundwork for us to support it.

**Comment by cup** [comment]
if you want something constructive, look at https://github.com/stedolan/jq/pull/1005

although I have to admit I am frustrated as that pull has been linked repeatedly already, in this thread.

**Comment by nicowilliams** [comment]
@cup please stop.

**Comment by cup** [comment]
@nicowilliams what is your problem? I linked to a pull that actually accomplishes (from personally testing) what he is asking for.
